packages/salesforce-adapter/src/filters/replace_instance_field_values.ts

```
@@ -0,0 +1,142 @@
+  let apiNameToId = await getIdsForType(client, CUSTOM_FIELD)
+  apiNameToId = Object.fromEntries(
+    Object.entries(apiNameToId).map(([name, id]) => [name, toShortId(id)])
+  )
+  return apiNameToId
+}
+
+const shouldReplace = (field: Field): boolean => {
+  const resolverFinder = generateReferenceResolverFinder(fieldSelectMapping)
+  return !_.isEmpty(resolverFinder(field))
+}
+
+const replaceInstanceValues = (instance: InstanceElement,
+  nameLookup: Record<string, string>): void => {
+  const transformFunc: TransformFunc = ({ value, field }) => {
+    if (_.isUndefined(field) || !shouldReplace(field)) {
+      return value
+    }
+
+    // if we can't find an item in the lookup it's because
+    // it's a standard field that doesn't need translation
+    return _.isArray(value)
+      ? value.map(s => nameLookup[s] ?? s)
+      : (nameLookup[value] ?? value)
+  }
+
+  const values = instance.value
+  instance.value = transformValues(
+    {
+      values,
+      type: instance.type,
+      transformFunc,
+      strict: false,
+    }
+  ) ?? values
+}
+
+const replaceInstancesValues = (elements: Element[], nameLookUp: Record<string, string>): void => {
+  elements
+    .filter(isInstanceElement)
+    .filter(e => Object.keys(metadataTypeToInstanceName).includes(metadataType(e)))
+    .forEach(e => {
+      replaceInstanceValues(e, nameLookUp)
+      log.debug(`replaced values of instance ${apiName(e)}`)
+    })
+}
+
+const getIdToNameLookupFromAllElements = (elements: Element[]): Record<string, string> => {
+  const lookup: Record<string, string> = {}
+  const fields = elements
+    .filter(isObjectType)
+    .filter(e => Object.values(metadataTypeToInstanceName).includes(apiName(e)))
+    .flatMap(e => Object.values(e.fields))
+
+  Object.assign(lookup, ...fields.map(field => {
+    const id = getInternalId(field)
+    return ({
+      [id === undefined ? id : toShortId(id)]: apiName(field),
```

might be better to exclude `undefined` ids from the map? in the past we've had some weird behaviors with a fallback to a random object because of that